### PR TITLE
Add llama 3.1 rope scaling factors to llama conversion and inference

### LINKED ISSUE #8650: Feature Request: Proper Llama 3.1 Support in llama.cpp
### Prerequisites

- [X] I am running the latest code. Mention the version if possible as well.
- [X] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [X] I searched using keywords relevant to my issue to make sure that I am creating a new issue that is not already open (or closed).
- [X] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new and useful enhancement to share.

### Feature Description

Llama 3.1 was just released and it is a significant leg up from the previous series of models: https://huggingface.co/blog/llama31

Whilst the overall architecture is the same, it requires some modelling updates, primarily around RoPE scaling: https://github.com/huggingface/transformers/blob/bc2adb0112b6677b0dfb4105c74570a0f92183eb/src/transformers/modeling_rope_utils.py#L298

It'd be great to add support for those so that the generations are more coherent and make sense.

### Motivation

Note: Without the modelling changes, the generation might look coherent, but they are far from great and the true-st potential of the model!

### Possible Implementation

Here's the corresponding transformers implementation: https://github.com/huggingface/transformers/blob/bc2adb0112b6677b0dfb4105c74570a0f92183eb/src/transformers/modeling_rope_utils.py#L298

### PR DESCRIPTION
**Pull request by jmorganca**
Hi all, this commit generates the rope factors on conversion and adds them to the resulting model as a tensor. At inference time, these factors are passed to the `ggml_rope_ext` rope operation.

From our testing, this really improves results for context windows above 8192 for Llama 3.1.

This should fix https://github.com/ggerganov/llama.cpp/issues/8650

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [ ] Low
  - [x] Medium
  - [ ] High

### CODE REVIEW

File: convert_hf_to_gguf.py
Comment (jmorganca) [inline convert_hf_to_gguf.py:1544]: Unlike the transformers or pytorch reference code, I believe we need to invert `rope_freqs` here, since they applied later in an inverted way in the rope kernels

File: convert_hf_to_gguf.py
Comment (Galunid) [inline convert_hf_to_gguf.py:1518]: I think it's best if you merge this into the rope scaling above (line 1501), so that all the rope operations are in one place.

**Comment by jmorganca** [comment]
Marking this as draft as there may also be a change required to the rope kernels – investigating!

File: convert_hf_to_gguf.py
Comment (jxy) [reply convert_hf_to_gguf.py:1544]: The `freq_factors` in ggml.c are divided into `theta`, which is $p b^{-2 i/d}$ https://github.com/ggerganov/llama.cpp/blob/68504f0970db5a3602d176953690f503059906b1/ggml/src/ggml-metal.metal#L1567-L1571 and then taken `cos` and `sin`. But the tensor here you created is $f / b^{-2 i/d}$. Perhaps the factors should be built with `factors.append(1)`, `factors.append(factor)`, `factors.append(1/((1-s)*factor+s))` depending on the wavelen?

**Comment by jxy** [comment]
Probably it's not a good idea to change the kernels. I just tested creating the factors instead as in my comment above.  Here's the changes on top of your PR,
```diff
diff --git a/convert_hf_to_gguf.py b/convert_hf_to_gguf.py
index 4422948f..897a6e33 100755
--- a/convert_hf_to_gguf.py
+++ b/convert_hf_to_gguf.py
@@ -1518,7 +1518,7 @@ class LlamaModel(Model):
             if rope_scaling.get("rope_type", '').lower() == "llama3":
                 base = hparams.get("rope_theta", 10000.0)
                 dim = int((hparams["hidden_size"] // hparams["num_attention_heads"]) * hparams.get("partial_rotary_embeddings", 1.0))
-                inv_freq = 1.0 / (base ** (torch.arange(0, dim, 2, dtype=torch.float32) / dim))
+                freqs = 1.0 / (base ** (torch.arange(0, dim, 2, dtype=torch.float32) / dim))
 
                 factor = rope_scaling.get("factor", 8.0)
                 low_freq_factor = rope_scaling.get("low_freq_factor", 1.0)
@@ -1528,20 +1528,20 @@ class LlamaModel(Model):
                 low_freq_wavelen = old_context_len / low_freq_factor
                 high_freq_wavelen = old_context_len / high_freq_factor
 
-                rope_freqs = []
-                for freq in inv_freq:
+                rope_factors = []
+                for freq in freqs:
                     wavelen = 2 * math.pi / freq
                     if wavelen < high_freq_wavelen:
-                        rope_freqs.append(freq)
+                        rope_factors.append(1)
                     elif wavelen > low_freq_wavelen:
-                        rope_freqs.append(freq / factor)
+                        rope_factors.append(factor)
                     else:
                         assert low_freq_wavelen != high_freq_wavelen
                         smooth = (old_context_len / wavelen - low_freq_factor) / (high_freq_factor - low_freq_factor)
-                        rope_freqs.append((1 - smooth) * freq / factor + smooth * freq)
+                        rope_factors.append(1 / ((1 - smooth) / factor + smooth))
 
                 self.gguf_writer.add_rope_scaling_attn_factors(1.0)
-                self.gguf_writer.add_tensor(gguf.TENSOR_NAMES[gguf.MODEL_TENSOR.ROPE_FREQS] + ".weight", 1.0 / np.array(rope_freqs))
+                self.gguf_writer.add_tensor(gguf.TENSOR_NAMES[gguf.MODEL_TENSOR.ROPE_FREQS] + ".weight", np.array(rope_factors, dtype = np.float32))
 
     @staticmethod
     def permute(weights: Tensor, n_head: int, n_head_kv: int | None):
```

A quick test shows perplexity is slightly better than master branch ([1]4.6850,[2]5.9231),
```
perplexity: calculating perplexity over 17 chunks, n_ctx=16384, batch_size=2048, n_seq=1
perplexity: 191.29 seconds per pass - ETA 54.18 minutes
[1]4.3272,[2]5.5469,^C
```

File: convert_hf_to_gguf.py
Comment (jxy) [inline convert_hf_to_gguf.py:1541]: We want the invert of https://github.com/meta-llama/llama-models/blob/1b5892739868e5333fb7f022ba91218f0ae5f9c2/models/llama3_1/api/model.py#L62 factoring out the `freq`, which should be `1/((1 - smooth) / scale_factor + smooth)`

**Comment by jxy** [comment]
It should be
```
1 / ((1 - smooth) / factor + smooth)
```

**Comment by jmorganca** [comment]
Awesome, thanks @jxy. Should be updated now. Agreed much better without the kernel changes.

File: convert_hf_to_gguf.py
Comment (compilade) [inline convert_hf_to_gguf.py:1544]:
```suggestion
                self.gguf_writer.add_tensor(self.format_tensor_name(gguf.MODEL_TENSOR.ROPE_FREQS), np.array(rope_factors, dtype=np.float32))
```

This should be equivalent, but it has additional sanity checks.

**Comment by kallewoof** [comment]
Nice. Tested doing a bunch of summaries using up the entire 128k context and the output looks good whereas on master it outputs broken garbage.

**Comment by MoonRide303** [comment]
I've notice one potential problem - GGUF made with this PR has n_tensors = 292 instead of 291 on master (extra rope_freqs.weight). It might cause errors in some apps checking it (like koboldcpp):
```
llama_model_load: error loading model: done_getting_tensors: wrong number of tensors; expected 292, got 291
llama_load_model_from_file: failed to load model
Traceback (most recent call last):
  File "koboldcpp.py", line 4208, in <module>
    main(parser.parse_args(),start_server=True)
  File "koboldcpp.py", line 3883, in main
    loadok = load_model(modelname)
  File "koboldcpp.py", line 773, in load_model
    ret = handle.load_model(inputs)
OSError: exception: access violation reading 0x0000000000001894
[21452] Failed to execute script 'koboldcpp' due to unhandled exception!
```

FYI @LostRuins

**Comment by tristandruyen** [comment]
> Initial tests show improvement over master (model able to reason within bigger context - I tested up to `-c 32768`), but I've noticed one potential problem - GGUF made with this PR has n_tensors = 292 instead of 291 on master (extra rope_freqs.weight). It might cause errors in some apps checking it (like [koboldcpp](https://github.com/LostRuins/koboldcpp)):
> 

Isn't this mismatch only due to koboldcpp not being updated yet ?

**Comment by tristandruyen** [comment]
> Initial tests show improvement over master (model able to reason within bigger context - I tested up to `-c 32768`), but I've noticed one potential problem - GGUF made with this PR has n_tensors = 292 instead of 291 on master (extra rope_freqs.weight). It might cause errors in some apps checking it (like [koboldcpp](https://github.com/LostRuins/koboldcpp)):
> 

Isn't this just a side-effect of kobold.cpp using an outdated llama.cpp version internally ?

**Comment by MoonRide303** [comment]
> > Initial tests show improvement over master (model able to reason within bigger context - I tested up to `-c 32768`), but I've noticed one potential problem - GGUF made with this PR has n_tensors = 292 instead of 291 on master (extra rope_freqs.weight). It might cause errors in some apps checking it (like [koboldcpp](https://github.com/LostRuins/koboldcpp)):
> 
> Isn't this just a side-effect of kobold.cpp using a different llama.cpp version internally, which would need to be updated anyways to take advantage of these changes ?

I tested it on koboldcpp 1.71 (released today, having merged upstream changes yesterday) - but you might be right some changes from original llama.cpp are missing. I am not sure if it's something to worry about on the llama.cpp project side, just pointing it out as potential side effect - for some people new GGUFs might stop working.

**Comment by LostRuins** [comment]
I wasn't aware that this PR breaks previously quantized Llama 3.1 models, I thought that all prior models would continue to work. 
Shouldn't the tensor count always match what's declared in the metadata regardless?

Anyway, I'll do a patch release once this is merged into master.

**Comment by LostRuins** [comment]
Yeah but that's my point - it's failing to load because the tensor counts don't match the metadata, which it should in both cases, unless I misunderstand the error.

**Comment by kallewoof** [comment]
Merging with koboldcpp concedo_experimental works fine on my end, FWIW.

**Comment by schmorp** [comment]
Why couldn't this tensor be added by llama.cpp when loading? superficially it doesn't make much sense to bake the rope config into the model at conversion time, and prevents bugfixes at a later time.

**Comment by jmorganca** [comment]
Thanks! Given the hard check on the tensor count at load time, it seems we might need to instead add the rope parameters as metadata and calculate it at runtime. Definitely don't want to break backwards compatibility with old GGUF files with this change sorry

**Comment by ggerganov** [comment]
Thank you for looking into this. Btw even if it is in the metadata and the rope freqs are computed at runtime, it will still not be backwards compatible because the old GGUF files will not have the metadata

**Comment by 3Simplex** [comment]
Thank you for your effort on this, we are all waiting patiently for the rope to be finished! We have some great functionality in GPT4All with this new model!

image

**Comment by jmorganca** [comment]
@ggerganov do you know if there a preferred way to create tensors from scalar values (vs using `ml.create_tensor`)? Working on an iteration of this PR that does this (vs setting it as a tensor in the GGUF file)

**Comment by gilbertgong** [comment]
I tried testing this (building off this branch and doing a fresh Q8 gguf with the resulting convert script), in comparison to simply using `--rope-freq-base 8000000` with the unpatched binary and just the change to `llama-bpe` (as suggested in the linked thread), and actually simply using `--rope-freq-base 8000000` seemed to give overall higher quality from a subjective perspective.  Have others had good results with this patch?

**Comment by ggerganov** [comment]
@jmorganca Creation of tensors at runtime hasn't been needed in the past, so we haven't considered it yet. I think it would require to create additional contexts for runtime-computed tensors for each layer (see `llama_kv_cache_init` for something similar). Then the runtime data has to be computed on the CPU and applied to the tensors with `ggml_backend_tensor_set`.

IMO it would be much simpler to keep the frequency computation logic in the Python code and proceed with the current PR as it is (+ addressing @Galunid and @compilade comments). This way it is also more inline with the Phi3 long/short rope frequency logic. I don't think there are significant advantages to computing the rope freqs at runtime from the metadata and moreover both approaches are backwards incompatible so new GGUFs will have to be created either way.

> Given the hard check on the tensor count at load time

Not sure which logic is that, can you clarify?

**Comment by jmorganca** [comment]
@ggerganov okay, thanks. I can address the comments to unblock the PR

This is the check https://github.com/ggerganov/llama.cpp/blob/master/src/llama.cpp#L4142 that seems to cause the error:

```
llama_model_load: error loading model: done_getting_tensors: wrong number of tensors; expected 292, got 291
```

Do you know if there's a way we might be able to work around that somehow?

**Comment by ggerganov** [comment]
> The thinking around computing them at runtime is old clients will continue to run the new models (albeit not well with n_ctx > 8192)

The proposed changes to the RoPE also affect `n_ctx < 8192`. So even if old clients were able to use the new models, the generations would be technically incorrect even for shorter contexts, because of the missing factors. Would this be desired?

But regardless, I can't think of a simple workaround that would enable this

**Comment by m18coppola** [comment]
I'd rather see the old llama-3.1 GGUF's totally break rather than risking users downloading an old model with degraded quality and giving them a bad impression. A program crash will queue the user to investigate what's wrong. Degraded performance just might make the user think the software/model is of low quality.

**Comment by qnixsynapse** [comment]
>  A program crash will queue the user to investigate what's wrong. Degraded performance just might make the user think the software/model is of low quality.

Exactly! I see no reason to make it backward compatible.

**Comment by LostRuins** [comment]
Crashing outright is bad, a better result would be a warning printed to stdout.

sort of like how we did this:
https://github.com/ggerganov/llama.cpp/blob/01245f5b1629075543bc4478418c7d72a0b4b3c7/src/llama.cpp#L5344-L5352

**Comment by slaren** [comment]
It doesn't crash, it causes `llama_load_model_from_file` to return an error.

File: convert_hf_to_gguf.py
Comment (bartowski1182) [inline convert_hf_to_gguf.py:1595]: should we assert if they're even or just not apply it at all and return gracefully?

**Comment by gilbertgong** [comment]
Another way to make it backwards compatible if desired is to just use the old rope calculation if the metadata/tensor is missing.

File: convert_hf_to_gguf.py
Comment (compilade) [inline convert_hf_to_gguf.py:1599]:
```suggestion
                self.gguf_writer.add_tensor(self.format_tensor_name(gguf.MODEL_TENSOR.ROPE_FREQS), np.array(rope_factors, dtype=np.float32))
```

Same comment as before, see <https://github.com/ggerganov/llama.cpp/pull/8676#discussion_r1690771131>. Not sure why this was reverted in <https://github.com/ggerganov/llama.cpp/pull/8676/commits/a946b40eee06b2eea307235d5c1138427e75d9bd>.

File: convert_hf_to_gguf.py
Comment (compilade) [reply convert_hf_to_gguf.py:1595]: I think this assertion should be outside of the loop. These values don't change within the loop. If it's gracefully handled, it should at least print a warning with `logger.warning("message here")`.

**Comment by m18coppola** [comment]
> Another way to make it backwards compatible if desired is to just use the old rope calculation if the metadata/tensor is missing.

If so, I think there should be a warning similar to what @LostRuins mentioned.

**Comment by compilade** [comment]
> Another way to make it backwards compatible if desired is to just use the old rope calculation if the metadata/tensor is missing.

That's what the code already seems to do. It only uses the rope freqs when present.

> If so, I think there should be a warning 

I'm not sure there *can* be such a warning, because older models like `llama-2` still use the same model architecture (`LLM_ARCH_LLAMA`, so internally they can't really be distinguished), yet they don't require to use these modified rope freqs.

File: convert_hf_to_gguf.py
Comment (jmorganca) [reply convert_hf_to_gguf.py:1599]: Shoot, thanks. Not sure what happened there sorry.

**Comment by ddh0** [comment]
Just cloned branch `jmorganca:master`as of commit `90fd87df4155aef5f099812a99c1e06c0b588c0d` and used it to convert llama 3.1 8B instruct model and quantize to q8_0. Everything is working perfectly as far as I can tell. The only thing that I see potentially wrong is the pre-tokenizer is reported as `smaug-bpe`. Not sure if that's affecting anything or not.

File: src/llama.cpp
Comment (compilade) [inline src/llama.cpp:6058]:
```suggestion
                        layer.rope_freqs = ml.create_tensor(ctx_layer, tn(LLM_TENSOR_ROPE_FREQS, "weight"), {n_embd/n_head/2}, llama_model_loader::TENSOR_NOT_REQUIRED | (i != 0 ? llama_model_loader::TENSOR_DUPLICATED : 0));
```

This doesn't seem vertically aligned with anything, so the extra spaces are not necessary.

File: convert_hf_to_gguf.py
Comment (compilade) [inline convert_hf_to_gguf.py:1576]: Where does `partial_rotary_embeddings` come from? I don't see it in Llama-3.1's `config.json`, and if present, it would mess with the rope_freqs tensor size in a way which is not handled when loading it from `src/llama.cpp`.

```suggestion
                dim = self.hparams["hidden_size"] // self.hparams["num_attention_heads"]
```

**Review by compilade — APPROVED** [review]

**Comment by compilade** [comment]
> The only thing that I see potentially wrong is the pre-tokenizer is reported as `smaug-bpe`. Not sure if that's affecting anything or not.

@ddh0 When I convert the latest Llama-3.1-8B-Instruct from <https://huggingface.co/meta-llama/Meta-Llama-3.1-8B-Instruct> (at HF commit [`07eb05b21`](https://huggingface.co/meta-llama/Meta-Llama-3.1-8B-Instruct/commit/07eb05b21d191a58c577b4a45982fe0c049d0693)) with the changes from this PR (at <https://github.com/ggerganov/llama.cpp/pull/8676/commits/90fd87df4155aef5f099812a99c1e06c0b588c0d>), the pre-tokenizer is `llama-bpe`, so I guess the pre-tokenizer differences have been fixed upstream.

```console
$ sha256sum Meta-Llama-3.1-8B-Instruct/tokenizer{,_config}.json 
79e3e522635f3171300913bb421464a87de6222182a0570b9b2ccba2a964b2b4  Meta-Llama-3.1-8B-Instruct/tokenizer.json
24e8a6dc2547164b7002e3125f10b415105644fcf02bf9ad8b674c87b1eaaed6  Meta-Llama-3.1-8B-Instruct/tokenizer_config.json
```

**Comment by ddh0** [comment]
Thanks @compilade, you were right. After pulling the latest repo the tokenizer is detected as `llama-bpe`. Thanks

File: convert_hf_to_gguf.py
Comment (jmorganca) [reply convert_hf_to_gguf.py:1576]: Good catch! Sorry!

**Pull request merged**
